pathology/hovernet/training_ignite.py

```diff
@@ -0,0 +1,303 @@
+    return log_dir
+
+def set_device(cfg):
+    return torch.device(cfg['gpu'] if torch.cuda.is_available() else "cpu")
+
+def prepare_datasets(data_dir):
+    info = pd.read_csv(os.path.join(data_dir, "patch_info.csv"))
+    file_names = np.squeeze(info.to_numpy()).tolist()
+
+    img_sources = [v.split("-")[0] for v in file_names]
+    img_sources = np.unique(img_sources)
+
+    cohort_sources = [v.split("_")[0] for v in img_sources]
+    _, cohort_sources = np.unique(cohort_sources, return_inverse=True)
+
+    splitter = StratifiedShuffleSplit(n_splits=1, train_size=0.8, test_size=0.2, random_state=0)
+
+    split_generator = splitter.split(img_sources, cohort_sources)
+    for train_indices, valid_indices in split_generator:
+        train_cohorts = img_sources[train_indices]
+        valid_cohorts = img_sources[valid_indices]
+        if np.intersect1d(train_cohorts, valid_cohorts).size != 0:
+            raise ValueError("Train and validation cohorts has an overlap.")
+        train_names = [
+            file_name for file_name in file_names for source in train_cohorts if source == file_name.split("-")[0]
+        ]
+        valid_names = [
+            file_name for file_name in file_names for source in valid_cohorts if source == file_name.split("-")[0]
+        ]
+        train_names = np.unique(train_names)
+        valid_names = np.unique(valid_names)
+        print(f"Train: {len(train_names):04d} - Valid: {len(valid_names):04d}")
+        if np.intersect1d(train_names, valid_names).size != 0:
+            raise ValueError("Train and validation cohorts has an overlap.")
+
+        train_indices = [file_names.index(v) for v in train_names]
+        valid_indices = [file_names.index(v) for v in valid_names]
+
+    images = np.load(os.path.join(data_dir, "images.npy"))
+    labels = np.load(os.path.join(data_dir, "labels.npy"))
+
+    data = [
+        {
+            "image": image,
+            "image_meta_dict": {"original_channel_dim": -1},
+            "label": label,
+            "label_meta_dict": {"original_channel_dim": -1},
+        }
+        for image, label in zip(images, labels)
+    ]
+
+    train_data = [data[i] for i in train_indices]
+    valid_data = [data[i] for i in valid_indices]
+
+    return train_data, valid_data
+
+def cropping_center(x, crop_shape, batch=False):
+    """Crop an input image at the centre.
+    Args:
+        x: input array
+        crop_shape: dimensions of cropped array
+
+    Returns:
+        x: cropped array
+
+    """
+    orig_shape = x.shape
+    if not batch:
+        h0 = int((orig_shape[1] - crop_shape[0]) * 0.5)
+        w0 = int((orig_shape[2] - crop_shape[1]) * 0.5)
+        x = x[:, h0: h0 + crop_shape[0], w0: w0 + crop_shape[1]]
+    else:
+        h0 = int((orig_shape[2] - crop_shape[0]) * 0.5)
+        w0 = int((orig_shape[3] - crop_shape[1]) * 0.5)
+        x = x[..., h0: h0 + crop_shape[0], w0: w0 + crop_shape[1]]
+    return x
+
+def val_post_process(data):
```

Comment:
Should we add all the complex post-processing here before computing metrics?
I mean the transforms in PR: https://github.com/Project-MONAI/MONAI/pull/5173.

Thanks.

Reply:
I have just checked the original repo, you are exactly right, I will add it in the latest commit, thanks!